File: native/macos-vision-ocr.m
#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>
#import <Vision/Vision.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc < 2) {
            fprintf(stderr, "缺少图片路径\n");
            return 2;
        }

        NSString *imagePath = [NSString stringWithUTF8String:argv[1]];
        NSURL *imageURL = [NSURL fileURLWithPath:imagePath];
        NSImage *image = [[NSImage alloc] initWithContentsOfURL:imageURL];
        if (!image) {
            fprintf(stderr, "无法读取图片\n");
            return 3;
        }

        VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] init];
        request.revision = VNRecognizeTextRequestRevision3;
        request.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
        request.recognitionLanguages = @[@"zh-Hans", @"en-US"];
        request.usesLanguageCorrection = YES;

        NSError *error = nil;
        VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithURL:imageURL options:@{}];
        if (![handler performRequests:@[request] error:&error]) {
            fprintf(stderr, "文字识别失败：%s\n", error.localizedDescription.UTF8String);
            return 5;
        }

        NSArray<VNRecognizedTextObservation *> *observations = [request.results sortedArrayUsingComparator:^NSComparisonResult(VNRecognizedTextObservation *left, VNRecognizedTextObservation *right) {
            CGFloat verticalDelta = CGRectGetMidY(left.boundingBox) - CGRectGetMidY(right.boundingBox);
            if (fabs(verticalDelta) > 0.02) return verticalDelta > 0 ? NSOrderedAscending : NSOrderedDescending;
            if (CGRectGetMinX(left.boundingBox) < CGRectGetMinX(right.boundingBox)) return NSOrderedAscending;
            return NSOrderedDescending;
        }];

        NSMutableArray<NSString *> *lines = [NSMutableArray array];
        for (VNRecognizedTextObservation *observation in observations) {
            VNRecognizedText *candidate = [observation topCandidates:1].firstObject;
            NSString *line = [candidate.string stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceAndNewlineCharacterSet];
            if (line.length > 0) [lines addObject:line];
        }
        printf("%s\n", [[lines componentsJoinedByString:@"\n"] UTF8String]);
    }
    return 0;
}

File: native/macos-vision-ocr.swift
import AppKit
import Foundation
import Vision

guard CommandLine.arguments.count > 1 else {
    fputs("缺少图片路径\n", stderr)
    exit(2)
}

let imageURL = URL(fileURLWithPath: CommandLine.arguments[1])
guard let image = NSImage(contentsOf: imageURL) else {
    fputs("无法读取图片\n", stderr)
    exit(3)
}

var proposedRect = NSRect(origin: .zero, size: image.size)
guard let cgImage = image.cgImage(forProposedRect: &proposedRect, context: nil, hints: nil) else {
    fputs("无法解析图片\n", stderr)
    exit(4)
}

let request = VNRecognizeTextRequest()
request.recognitionLevel = .accurate
request.recognitionLanguages = ["zh-Hans", "zh-Hant", "en-US"]
request.usesLanguageCorrection = true

do {
    try VNImageRequestHandler(cgImage: cgImage, options: [:]).perform([request])
    let lines = (request.results ?? [])
        .sorted { lhs, rhs in
            let verticalDelta = lhs.boundingBox.midY - rhs.boundingBox.midY
            if abs(verticalDelta) > 0.02 { return verticalDelta > 0 }
            return lhs.boundingBox.minX < rhs.boundingBox.minX
        }
        .compactMap { $0.topCandidates(1).first?.string }
        .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
    print(lines.joined(separator: "\n"))
} catch {
    fputs("文字识别失败：\(error.localizedDescription)\n", stderr)
    exit(5)
}
